packages/code-snippet/src/CodeSnippetWidget.tsx

```diff
@@ -214,37 +52,164 @@ export class CodeSnippetWidget extends ReactWidget {
   codeSnippetManager: CodeSnippetService;
   renderCodeSnippetsSignal: Signal<this, ICodeSnippet[]>;
   getCurrentWidget: () => Widget;
-
-  constructor(getCurrentWidget: () => Widget) {
+  editorFactory: CodeEditor.Factory;
+  app: JupyterFrontEnd;
+  restorer: ILayoutRestorer;
+  editorTracker: WidgetTracker<MetadataEditor>;
+
+  constructor(
+    getCurrentWidget: () => Widget,
+    editorFactory: CodeEditor.Factory,
+    app: JupyterFrontEnd,
+    restorer: ILayoutRestorer
+  ) {
     super();
     this.getCurrentWidget = getCurrentWidget;
     this.codeSnippetManager = new CodeSnippetService();
     this.renderCodeSnippetsSignal = new Signal<this, ICodeSnippet[]>(this);
+    this.openCodeSnippetEditor = this.openCodeSnippetEditor.bind(this);
+    this.saveSnippet = this.saveSnippet.bind(this);
+    this.editorFactory = editorFactory;
+    this.app = app;
+    this.app.commands.addCommand(METADATA_EDITOR_ID + ':open', {
+      isVisible: () => {
+        return false;
+      },
+      execute: (args: any) => {
+        this.openCodeSnippetEditor(
+          args.name,
+          args.description,
+          args.language,
+          args.code,
+          args.newFile
+        );
+      }
+    });
+    this.restorer = restorer;
+    this.editorTracker = new WidgetTracker<MetadataEditor>({
+      namespace: METADATA_EDITOR_ID
+    });
+    this.restorer.restore(this.editorTracker, {
+      name: (metadataEditor: any) => {
+        return metadataEditor.metadata.name;
+      },
+      command: METADATA_EDITOR_ID + ':open',
+      args: (widget: MetadataEditor) => ({
+        name: widget.metadata.name,
+        description: widget.metadata.description,
+        language: widget.metadata.language,
+        code: widget.metadata.code,
+        newFile: widget.newFile
+      })
+    });
+    this.fetchData = this.fetchData.bind(this);
+    this.updateSnippets = this.updateSnippets.bind(this);
   }
 
   // Request code snippets from server
   async fetchData(): Promise<ICodeSnippet[]> {
     return await this.codeSnippetManager.findAll();
   }
 
-  // Triggered when the widget button on side palette is clicked
-  onAfterShow(msg: Message): void {
+  updateSnippets(): void {
     this.fetchData().then((codeSnippets: ICodeSnippet[]) => {
       this.renderCodeSnippetsSignal.emit(codeSnippets);
     });
   }
 
+  // Triggered when the widget button on side panel is clicked
+  onAfterShow(msg: Message): void {
+    this.updateSnippets();
+  }
+
+  addCodeSnippet(): void {
+    this.openCodeSnippetEditor('', '', '', '', true);
+  }
+
+  saveSnippet(snippetEditor: MetadataEditor): void {
+    const newSnippet = {
+      schema_name: 'code-snippet',
+      name: snippetEditor.metadata.name,
```

Comment:
also, the `name` field is more restrictive than `display_name`. for example, you cannot have a space in the `name` field but it is acceptable in `display_name`. may want to think about normalizing the the `name` field (i.e., replace spaces with underscoring or removing unacceptable characters). the user does not necessarily see the `name` but they do see the `display_name` so keeping the `display_name` as they entered but modifying the `name` should suffice

Reply:
I agree that this makes sense to check on the client.  FYI - The server will apply the schema file against the sent payload to validate all fields (including required) anyway. 

For now, it's probably fine to use "hard-coded" required values and forgo application of the schema.

Reply:
@kevin-bates the more i think about it,  am i correct in my interpretation of the `name` field as sort of a unique ID field for the metadata?

if so, given that assumption, would it be worth allowing submission to the server without a `name` field but include `display_name` (i.e., you need a `name` and/or a `display_name`)? in cases where only the `display_name` is provided the server would generate a `name` (e.g., normalizing the `display_name` or generating a unique id).  

and this would be only applicable for creating not editing. since you shouldn't be allowed to edit the ID/name but you still able to edit the `display_name`.

that way
- the client doesn't have to be cognizant of underlining ID/name mechanism
- the end user just sees/creates the `display_name` without having to think about or make sure they are abiding by the underlining ID/name requirements
- if  ID/name validation requirements ever changed it can all be handled in server rather than having to make updates in server and all clients calling the POST API.

your thoughts? maybe a discussion for a later time...

Reply:
Hi @vabarbosa - this all makes sense.  I agree that the name displayed here should only be the display name.

Yes, the `name` is actually the filename of the persisted metadata - so in that sense - it is unique.  Keep in mind though, that _this_ uniqueness spans the hierarchy and is essentially how we allow factory data to be "updated" (although that update equates to a file's creation in another directory).

Looking at how we [currently deal with `name`](https://github.com/elyra-ai/elyra/blob/master/elyra/metadata/metadata.py#L266-L272) we also support hyphens and already enforce lowercase, so preserving hyphens, replacing spaces with underscores and lowercasing characters will be compatible.  We'd then just need to figure out what to do with 8-bit+ characters (and I'm guessing we'd just strip them).  And of course, we'd need to tolerate it being `None`.